### ISSUE TITLE
Prevent command spoofing

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#17135]
### Summary of the new feature / enhancement

Don't blindly overwrite commands that are part of a signed script or module.

Reading this [Colibri Loader combines Task Scheduler and PowerShell in clever persistence technique](https://blog.malwarebytes.com/threat-intelligence/2022/04/colibri-loader-combines-task-scheduler-and-powershell-in-clever-persistence-technique/), I am quiet surprised to see how easily it appears to overwrite a native (presumably signed) command.

Somehow, I think it should be possible to get the user's attention when this happens by asking whether to overwrite a function that is part of a signed script or module (when the `Set-ExecutionPolicy` is set to `RemoteSigned` or higher?):

### Proposed technical implementation details (optional)

Wishful thinking:

```PowerShell
PS C:\> Get-Content .\MyScript.ps1
function MySigned {
    Write-Host 'Hello World'
}
PS C:\> $Certificate = Get-ChildItem Cert:\LocalMachine\My | Where-Object {$_.Subject -eq 'CN=SelfSigned'}
PS C:\> Set-AuthenticodeSignature -FilePath MyScript.ps1 -Certificate $Certificate

PS C:\> . .\MyScript.ps1
PS C:\> MySigned
Hello World
PS C:\> function MySigned { Write-Host 'Hello Colibri' }
```
At this point I would like see a message in the direction of:

> Do you want to overwrite the command `MySigned` which is a part
> of the signed script `MyScript` from the publisher "SelfSigned"?
>
> [N] No  [Y] Yes [?] Help (default is "N"):

This would probably require that all functions (commands) that are part of a signed script or module should be securely marked as "signed" to be able to check for any possible unwanted overwrites.

**Notes**
* The question should also appear when the (signed) command is overwritten in a higher scope.
* The question should **not** appear when the command comes from the same source (script/module, aka publisher) as the existing command, e.g. in case of a recurse function.

**Comment by jhoneill** [comment]
It's a nice idea but 
1. Why would it apply only to functions - surely you would want it to apply to compiled cmdlets as well ? 
2. If it applied to built in cmdlets that would break existing Proxy functions.  For example Octopus deploy redefines `Write-Verbose` to capture verbose output to its own log.  `Write-verbose` is one of PowerShell's modules which is signed, so if this was a blanket block, octopus would fail
3. You would also need to prevent an alias being defined with the name of the function you want to protect. 
4. With a compiled cmdlet or executable file (including a .ps1) you can ensure the command that you run has not been subverted  by specifying its full path.  e.g. `Microsoft.PowerShell.Utility\Write-Verbose "boo" -Verbose`  this defeats aliases or proxy functions or a file higher up the search path running instead of the one you intended, but not redefining a function over the top of an existing one or creating an alias for a trusted function. 
5. To be used in an attack the function re-definition needs run between loading the legitimate function and running it. If you could do that, would you not just run the code which does the malicious thing, not set it as a function to be run if the  user runs some command later. ?    
6. When you're loading something where nothing else can be  allowed to sneak in (JEA scenarios being one) a combination of command visibility and using `$ExecutionContext.SessionState to change the language mode and the apps and scripts that can run generally does the job 
In this case I would make 
```
$ExecutionContext.SessionState.LanguageMode = [System.Management.Automation.PSLanguageMode]::RestrictedLanguage
```
The last line of the file.

**Comment by jhoneill** [comment]
The explanation at link has one oddity in it. Get-Variable.Exe gets run. Normally if there is a cmdlet and an .exe the cmdlet wins. So how is the exe getting run ?

**Comment by vexx32** [comment]
Yeah that's the most interesting part to me, that feels like it might be a bug. It's not clear from the article how / why exactly PowerShell is apparently running this executable when it starts, since from the article it doesn't appear any commands are actually passed to the PowerShell instance itself.

**Comment by jborean93** [comment]
It's not pretty but if you want to always run a cmdlet from a specific module without being overwritten you can do

```powershell
&(Get-Command -Name Write-Verbose -Module Microsoft.PowerShell.Utility) -Message abc
```

Using the FQDN cmdlet name `Microsoft.PowerShell.Utility\Write-Verbose` is also an option but keep in mind someone may define a new alias or function with that name overloading it. This is why `Get-Command` is better as it ensures it finds the comand in the module specified.

I believe Pester does something like this to store the set of "safe" commands in a hashtable rather than relying on the environment. It's not pretty but it works today.

**Comment by jhoneill** [comment]
> Using the FQDN cmdlet name `Microsoft.PowerShell.Utility\Write-Verbose` is also an option but keep in mind someone may define a new alias or function with that name overloading it. 
 
That's not quite right.  The fully specified name will override the function/alias.

**Comment by jborean93** [comment]
```powershell
Microsoft.PowerShell.Utility\Write-Output "bar"

Function Microsoft.PowerShell.Utility\Write-Output {
    "foo"
}

Microsoft.PowerShell.Utility\Write-Output "bar"
```

Outputs `bar` then `foo` as there's now a function with the name of the FQ cmdlet name.

**Comment by jhoneill** [comment]
> ```powershell
> Microsoft.PowerShell.Utility\Write-Output "bar"
> 
> Function Microsoft.PowerShell.Utility\Write-Output {
>     "foo"
> }
> 
> Microsoft.PowerShell.Utility\Write-Output "bar"
> ```
> 
> Outputs `bar` then `foo` as there's now a function with the name of the FQ cmdlet name.


OK I was assuming the function name was simply "Write-Output".  I should have read what you said more closely , you did say that a function could be created using the fully qualified name, and that _is_ correct.

**Comment by iRon7** [comment]
@jhoneill,

> (5) ... If you can run arbitrary code in the session where the script is going to run, why not just run your malicious code ...

I find founding security precautions always difficult: at one hand, you might be wrong and take away all the arguments for the actual proposal. At the other hand, you might be right and open de box of Pandora...

I am drifting away from the "Colibri loader" suggestion, but I think that command spoofing could potentially be used to get secure information and to get higher privileges where the _user account_ is already compromised (and the execution policy is not restricted) by piggybacking on procedures that normally require user intervention for security reasons, e.g.:
* by spoofing the `Read-Host -AsSecureString` cmdlet and capture secrets
* by spoofing any cmdlet (e.g. `Write-Host`) and act when `[Security.Principal.WindowsPrincipal] [Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole] 'Administrator')` equals `$True`

**Comment by jhoneill** [comment]
@iRon7  all valid points :-) 
In my profile I redefine CD (it's push-location with improved completers, validators and transformers), h (better selection from history) and have a proxy function which adds -Recurse to Select-String. 
For JEA I wrote proxy functions which restricted what cmdlets would do, even though they were running under a privileged account 

But any of these could be done maliciously - how many admins run PowerShell elevated with their default profile? If I can access your home directory and add something to your profile then... game over. 

I do have an idea how anti-spoofing might done from a script, but it would need some tools to make it workable. 
`Get-AuthenticodeSignature (get-command Write-Verbose).module.path | % signerCertificate  `

Will show you who signed the code for the present version of Write-Verbose. 
If you had a #requires directive which had a command name, and certificate parameters you could say 
"Abort the script if Read-Host doesn't come from the module with the right signature". 

But people would need to do work to use it . Something which blocked anything which puts in a substitute  for a command would break too many things.

**Comment by doctordns** [comment]
one small thing @jhoneill - in the old days, you had to alias cd to push-location to be able to go "back". Now, with PowerShell 7, this is built in so no need to re-alias.

**Comment by jhoneill** [comment]
@iRon7  - "With all respect" sounds like you're about to rudely disagree, but in fact we mostly agree, and I haven't found you rude, yet :-)  

My view is  

1. Changing what a command name does is a fundamental feature of PowerShell - whether that is the `sort` alias replacing `sort.com` on Windows, or a proxy function changing the `Add-ADGroupMember`  in a JEA environment to reject certain groups, or a custom profile adding recurse to `Select-String.`   The benefits are substantial and it has been used for a long time so even if that could be changed it would be undesirable to do so. 
2.  Like many things (including PowerShell itself)  attackers can abuse this feature.  If they are in a position to do so, they are already able to do malicious things (change trusted code, run arbitrary code in a user's session) and the _focus_ should be on preventing them having that access, mitigations if they do have that access are _secondary_ . I probably sound like _secondary_ equates to not worth bothering with :-)  Modifying a profile is just one case of this. I'll comment to the other issue on that.
3. There are multiple reasons for wanting to ensure that a specific version of a command is run. For example I want to ensure that my Proxy functions for JEA are being called not the unrestricted cmdlets, or I want to ensure that version X of a module is loaded which implements a feature, and security There isn't a one size fits all for this - but trusted code should be signed, and specifying that a script requires commands to have signatures that meet some criteria looks like a possible solution.

**Comment by jhoneill** [comment]
> @jhoneill,
> 
> > "With all respect"
> 
> Sorry, that's because I am not native English, I think the correct approach would have been: 

Your written English is better than some native speakers! :-)

**Comment by hisashishimoe** [comment]
:ok

Windows の メ`ル<https://go.microsoft.com/fwlink/?LinkId=550986> から送信

差出人: ***@***.***>
送信日r: 2022年4月14日 19:19
宛先: ***@***.***>
CC: ***@***.***>
件名: Re: [PowerShell/PowerShell] Prevent command spoofing (Issue #17135)


@jhoneill<https://github.com/jhoneill>,

"With all respect"

Sorry, that's because I am not native English, I think the correct approach would have been:

Your written English is better than some native speakers! :-)

―

**Comment by jhoneill** [comment]
> one small thing @jhoneill - in the old days, you had to alias cd to push-location to be able to go "back". Now, with PowerShell 7, this is built in so no need to re-alias.

@doctordns   I found out when I used cd - on something which didn't have my profile  - Changing the alias means one profile works on all versions of PowerShell 
And too many cd -  do this 

```
#650 PS7 ~\Documents\WindowsPowerShell> pwsh -noprofile
PowerShell 7.2.2
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.

PS C:\Users\mcp\Documents\WindowsPowerShell> cd ..
PS C:\Users\mcp\Documents> cd -
PS C:\Users\mcp\Documents\WindowsPowerShell> cd -    # Stack should be empty
PS C:\> cd -
Set-Location: There is no location history left to navigate backwards.
PS C:\>

PS C:\> exit
#651 PS7 ~\Documents\WindowsPowerShell> cd -
#652 PS7 ~\Documents\PowerShell> cd -   #empty stack does nothing
#653 PS7 ~\Documents\PowerShell> cd -
```
The stack isn't available with the newer implementation my customize version of cd lets me do cd = [tab] to peek down the stack without popping items off it, so I still need `push `instead of `set`. 
 (remember Novell supported cd ... and cd ....  I've got that too :-) )

**Comment by hisashishimoe** [comment]
Windows の メ`ル<https://go.microsoft.com/fwlink/?LinkId=550986> から送信

差出人: ***@***.***>
送信日r: 2022年4月14日 20:28
宛先: ***@***.***>
CC: ***@***.***>; ***@***.***>
件名: Re: [PowerShell/PowerShell] Prevent command spoofing (Issue #17135)


one small thing @jhoneill<https://github.com/jhoneill> - in the old days, you had to alias cd to push-location to be able to go "back". Now, with PowerShell 7, this is built in so no need to re-alias.

@doctordns<https://github.com/doctordns> I found out when I used cd - on something which didn't have my profile - Changing the alias means one profile works on all versions of PowerShell
And too many cd - do this

#650 PS7 ~\Documents\WindowsPowerShell> pwsh -noprofile

PowerShell 7.2.2

Copyright (c) Microsoft Corporation.



https://aka.ms/powershell

Type 'help' to get help.



PS C:\Users\mcp\Documents\WindowsPowerShell> cd ..

PS C:\Users\mcp\Documents> cd -

PS C:\Users\mcp\Documents\WindowsPowerShell> cd -    # Stack should be empty

PS C:\> cd -

Set-Location: There is no location history left to navigate backwards.

PS C:\>



PS C:\> exit

#651 PS7 ~\Documents\WindowsPowerShell> cd -

#652 PS7 ~\Documents\PowerShell> cd -   #empty stack does nothing

#653 PS7 ~\Documents\PowerShell> cd -

The stack isn't available with the newer implementation my customize version of cd lets me do cd = [tab] to peek down the stack without popping items off it, so I still need push instead of set.
(remember Novell supported cd ... and cd .... I've got that too :-) )

―

**Comment by SeeminglyScience** [comment]
Can anyone actually repro the same result of the article? As I would have expected I can't get it to work at all. Maybe it's missing a step?

**Comment by chuckrock0095** [comment]
WTF?

**Comment by iRon7** [comment]
> Can anyone actually repro the same result of the article? As I would have expected I can't get it to work at all. Maybe it's missing a step?

I think I am able to confirm the PowerShell part where indeed an `exe`cutable file replaces the PowerShell command.
The point is that there is an error in the description where <strike>`%APPDATA%\Local\Microsoft\WindowsApps `</strike> should be `%LOCALAPPDATA%\Microsoft\WindowsApps `:

```PowerShell
PS C:\> Copy-Item (Get-Command WhoAmi).Source "$Env:LOCALAPPDATA\Microsoft\WindowsApps\Get-Variable.exe"
PS C:\> Get-Variable
Computer\Username
```

**Comment by SeeminglyScience** [comment]
Yeah the Engine WG discussed this last night and @IISResetMe was able to repro.

We'd like the security WG to review this before we discuss it further. /cc @TravisEz13

**Comment by PaulHigin** [comment]
@WG-Security
It looks like this feature request is to prevent a function being created having the same name as a function exposed through a loaded signed module file.

PowerShell command precedence allows (proxy) functions and aliases to override module commands.  We recommend not changing this as it would be a significant breaking change.  In addition, users can already use module fully qualified command names to ensure they are running the command they expect, and this is encouraged as a best practice.

**Comment by iRon7** [comment]
@PaulHigin,

(I presume that with your comment, the issue is reviewed with security WG and I am discuss this further)

> It looks like this feature request is to prevent a function being created having the same name as a function exposed through a loaded signed module file.

I like to refine this to:

> Prevent a function being *unnoticeably* and *undeniably* created with the same name as a function exposed through a loaded signed module file.

Besides, I think that any corrective action towards a better security might simply be optional to avoid any significant break change to developers that care less.

**Comment by ThomasNieto** [comment]
@PaulHigin If the recommendation is to use fully-qualified commands to prevent spoofing. Would it be considered a bug that fully-qualified command can be overwritten with a function or alias?


```powershell
PS C:\> Microsoft.PowerShell.Utility\Write-Output 'output'
output
PS C:\> function Microsoft.PowerShell.Utility\Write-Output { 'spoofed via function' }
PS C:\> Microsoft.PowerShell.Utility\Write-Output 'output'
spoofed via function
PS C:\> function spoofed { 'spoofed via alias' }
PS C:\> New-Alias -Name Microsoft.PowerShell.Utility\Write-Output -Value spoofed
PS C:\> Microsoft.PowerShell.Utility\Write-Output 'output'
spoofed via alias
```

**Comment by PaulHigin** [comment]
Good question.  Yes, it seems to me that fully qualified names should resolve to modules.  Tagging for committee review.

**Comment by jborean93** [comment]
Just a small reminder, you can use `Get-Command` to ensure you get a command in a specific module as per https://github.com/PowerShell/PowerShell/issues/17135#issuecomment-1095461560.

**Comment by iRon7** [comment]
> Just a small reminder, you can use **`Microsoft.PowerShell.Core\Get-Command`** to ensure you get a command in a specific module as per [#17135 (comment)](https://github.com/PowerShell/PowerShell/issues/17135#issuecomment-1095461560).

🤔

If I understand this correctly, this means that if you sign a PowerShell script, you better use fully qualified names for every command you invoke in the concerned script?

If that is true, I think that it would be appropriate to warn the signer if friendly command names are used in the concerned script to be signed or even possibly (optionally) replace any friendly command name with a fully qualified name (returned by `Microsoft.PowerShell.Core\Get-Command`) automatically prior the actual signing is done.  
The cmdlet `Microsoft.PowerShell.Security\Set-AuthenticodeSignature` would probably be the best cmdlet to accomplish this.

**Comment by doctordns** [comment]
@iRon7 a key issue for me here is that you cannot just use what you feel to be a fully qualified function name to avoid this.

If you use:
```powershell
Function Microsoft.PowerShell.Utility\Write-Output {"foo"}
```
Then the name of this function is the same as a fully qualified name.

Should PowerShell allow a function to have a name that is the same as a fully qualified function name?

**Comment by IISResetMe** [comment]
> If I understand this correctly, this means that if you sign a PowerShell script, you better use fully qualified names for every command you invoke in the concerned script?

@iRon7 The purpose of script signing is to authenticate the exact contents of a file on disk, nothing more. It has nothing to do with command discovery at runtime. The fact that the source script is signed was never meant to provide the kinds of guarantees your appear to be expecting

**Comment by jhoneill** [comment]
> @iRon7 a key issue for me here is that you cannot just use what you feel to be a fully qualified function name to avoid this.
> 
> If you use:
> 
> ```powershell
> Function Microsoft.PowerShell.Utility\Write-Output {"foo"}
> ```
> 
> Then the name of this function is the same as a fully qualified name.
> 
> **Should PowerShell allow a function to have a name that is the same as a fully qualified function name?**

TLDR **NO**  And I don't think this should be limited to functions either. 

Until @jborean93  pointed it out, I would have expected that name to be illegal.     
Although technically forbidding slash and backslash in function (and alias) names would be a breaking change,  I can't think of a legitimate use for them.  **This** is another example that should not be possible, 
```
> function Notepad {"boo"} 
> notepad 
boo       # <-- expected , desirable even.
> New-Alias -Name  C:\windows\notepad.exe -Value foo         
> C:\windows\notepad.exe          
boo       # <-- Far from desirable. 
```
 `:` is illegal in windows file systems but is needed to have C:, D:  swap drives so is allowed in function and alias names.  

But is there any reason for allowing `\  `and `/ `?    - `function:\` and `function:/` are the root of the function PSDrive, since this drive doesn't have an kind of container I understand it doesn't need those characters to serve as path separators, but I can't find any logic to say they _should_  or _must_ be allowed in Alias or function names. I'm not sure that losing "." (or : from aliases) would be harmful either.

**Comment by iRon7** [comment]
FYI: I have just also formally reported the malicious intents that might result from this issue through [Secure@](https://github.com/PowerShell/PowerShell/issues/secure@microsoft.com) because it is unclear to me whether this has already been done (taking the fact that this issue started as a propose and later has been flagged with `WG-Security`).

**Comment by jhoneill** [comment]
> > If I understand this correctly, this means that if you sign a PowerShell script, you better use fully qualified names for every command you invoke in the concerned script?
> 
> @iRon7 The purpose of script signing is to authenticate the exact contents of a file on disk, nothing more. It has nothing to do with command discovery at runtime. The fact that the source script is signed was never meant to provide the kinds of guarantees your appear to be expecting

It is any easy mistake to assume that if you have signed a script then if the signature says the script is unchanged then what runs is the same every time the script runs. How can something unchanged run different code?  Many ways, as it turns out. 

One can create what looks like a new version of the module which contains some commonly called cmdlet... 
 
Unless the script is run in a clean session with -noprofile, and explicitly imports the modules it uses, first checking their signatures (I'll target `Get-AuthenticodeSignature` ) and verifies that every alias, function, cmdlet or external file called is the expected one (i.e. Get-Widget is a cmdlet in the verified Widget module) . The signature is just "James was last to touch this part of a bigger whole, and this part hasn't been changed" .

I've signed my own scripts and put them into production (so if I went bad - who knows what was signed), and signed other peoples : in theory I reviewed before signing, but who knows what I might have added because I don't think people checked the code was as they gave it to me. When an organization signs its scripts the value of the signature depends on the process around it.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and agree that this is not a security issue in general, but would like the WG-Security group to revisit specifically the fully qualified cmdlet path and then the WG-Engine group to discuss the impact if we decide to not allow function names to resemble a fully qualified path.  Also consider if proxy functions may be impacted.  Expectation is that users should be able to rely on the fully qualified cmdlet path to know they are calling that cmdlet.

**Comment by iRon7** [comment]
Thanks all, for the comments on this issue.
The initial *propose* (not a bug or a security report) was to give signed functions a higher privilege as I was under the impression that signing scripts would always require the users attention (and be traceable) but apparently that is not the case.
I am ok to close the issue in `Resolution-Won't Fix` or keep it open for the "*expectation that users should be able to rely on the fully qualified cmdlet path to know they are calling that cmdlet*".
Do you like me to change the title for this? or just close the issue? and create a new "fully qualified cmdlet path" issue?

**Comment by JamesWTruher** [comment]
The engine WG has discussed this again and are concerned that any change here would be a breaking change. We are concerned as well as to how this might affect Pester as we believe that it takes advantage of this behavior.

**Comment by ThomasNieto** [comment]
If that is the case what is the user friendly  recommended way to do this without some hacky workaround using Get-Command?

**Comment by daxian-dbw** [comment]
The [Mocking](https://pester-docs.netlify.app/docs/usage/mocking) feature in Pester **may** depend on defining a fully-qualified name as an alias to intercept the call to a module function.

https://github.com/pester/Pester/blob/ec2131a4640461f8f0b32c499ebce393019aa392/src/functions/Mock.ps1#L218-L220
https://github.com/pester/Pester/blob/ec2131a4640461f8f0b32c499ebce393019aa392/src/functions/Mock.ps1#L547-L714

![image](https://user-images.githubusercontent.com/127450/165864976-87584568-6dc3-4424-96e8-773768f1f2cc.png)

So, this breaking change could be a big impact.

Aside from the breaking change concern, the Engine WG would like to have the Security WG (/cc @PaulHigin and @TravisEz13) to review again and clarify whether the current behavior (allowing function/alias with fully-qualified name) is really a security issue. We believe it's not (we could be wrong) given that if a bad actor can plant a function to override a module function, then the same actor can create a module with the same name and trick the user to load that module instead, even if we disallow function/alias with fully-qualified name.

**Comment by iSazonov** [comment]
Can we think about strict mode/constrained language?

**Comment by ThomasNieto** [comment]
Regardless if this is a security issue or not. This is still a reliability problem. If the script or module author cannot ensure the fully qualified cmdlet being called has the same behavior in dev as in production. This can and will cause unexpected behavior and the end user will blame the script or module author rather than the person who overloaded the cmdlets with a custom function or alias.

**Comment by JamesWTruher** [comment]
@PowerShell/powershell-committee discussed on 5/22/2022
We believe this is behaving as designed and any change here would be a breaking behavior.
